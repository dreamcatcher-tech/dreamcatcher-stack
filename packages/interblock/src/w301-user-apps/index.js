export * as crm from './src/crm'
