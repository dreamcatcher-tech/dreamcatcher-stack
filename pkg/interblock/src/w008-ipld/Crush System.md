# The Crush System

This collection of classes are designed to break a Pulse down into IPFS compatible chunks, for efficient hashing.

## Requirements

1. use the inheritance of IpldInterface as identifying feature of classes that get crushed
1. crush relies on an unbroken chain of crushable classes from the root down
   1. ie: cannot have a pojo that holds a crush class in it and expect it to work
1. arrays of crush classes follow these rules
1. when flagged, some children may be 'punched out' as CID links, and reinflated back in
1. any tree of classes can be reinflated from ipfs back into classes
1. ipfs blocks representing the different between the previous 'crush' process can be fetched then persisted in ipfs storage
1. be able to directly represent an IPFS primitive as a class, such as a CID - these are needed to signal to the ORM that it should stop rehydration
1. Be able to connect a set of IPLD schemas to the classes, to check them for correctness on crush and uncrush
   1. This check should be disablable for speed
