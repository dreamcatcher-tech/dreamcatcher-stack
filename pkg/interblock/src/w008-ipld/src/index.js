export { Binary } from './Binary'
export { RequestId } from './RequestId'
export { Request } from './Request'
export { RxRequest } from './RxRequest'
export { Reply } from './Reply'
export { RxReply } from './RxReply'
export { Address } from './Address'
export { PulseLink, HistoricalPulseLink } from './PulseLink'
export { TxQueue } from './TxQueue'
export { Tx } from './Tx'
export { RxQueue } from './RxQueue'
export { Rx } from './Rx'
export { Channel } from './Channel'
export { Config } from './Config'
export { PublicKey } from './PublicKey'
export { Validators } from './Validators'
export { ChannelsHamt } from './ChannelsHamt'
export { ChildrenHamt } from './ChildrenHamt'
export { UplinksHamt } from './UplinksHamt'
export { DownlinksHamt } from './DownlinksHamt'
export { SymlinksHamt } from './SymlinksHamt'
export { AddressesHamt } from './AddressesHamt'
export { Channels } from './Channels'
export { Transmissions } from './Transmissions'
export { Network } from './Network'
export { AsyncRequest } from './AsyncRequest'
export { AsyncTrail } from './AsyncTrail'
export { Pending } from './Pending'
export { State } from './State'
export { Timestamp } from './Timestamp'
export { Dmz } from './Dmz'
export { Provenance } from './Provenance'
export { Interpulse } from './Interpulse'
export { Pulse } from './Pulse'
export { Keypair } from './Keypair'
export { decode } from './IpldUtils'
export { IpldInterface } from './IpldInterface'
export { Hamt } from './Hamt'
