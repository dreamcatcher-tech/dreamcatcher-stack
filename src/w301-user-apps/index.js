const { covenantIdModel } = require('../w015-models')

const { crm } = require('./src/crm')
crm.covenantId = covenantIdModel.create('crm')

module.exports = { crm }
