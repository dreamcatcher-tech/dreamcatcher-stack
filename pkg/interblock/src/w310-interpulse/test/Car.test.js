import { Interpulse } from '../..'
import { Pulse } from '../../w008-ipld'
import { crm } from '../../w301-user-apps'
import Debug from 'debug'

const debug = Debug('tests')

describe('Car', () => {
  test('export', async () => {
    const engine = await Interpulse.createCI({
      overloads: { '/crm': crm.covenant },
    })
    await engine.add('app', '/crm')
    debug('begin export')
    const car = await engine.export('/app')
    debug('export complete')

    const blank = await Interpulse.createCI()
    debug('import start')
    const { roots, count } = await blank.import(car)
    debug('import end')
    expect(count).toMatchSnapshot()
    expect(roots.length).toBe(1)
    const [imported] = roots
    expect(imported).toBeInstanceOf(Pulse)

    const result = await blank.insert(imported.cid.toString(), 'forked')
    const forkedCurrent = await blank.current('forked')
    expect(imported.cid.toString()).toBe(forkedCurrent.cid.toString())
    // TODO address should be different at the insertion ?
    const schedule = await blank.current('forked/schedules')
    expect(schedule).toBeInstanceOf(Pulse)
    const originalPulseId = schedule.toString()
    debug('originalPulseId', originalPulseId)

    debug('begin write to fork')
    await blank.ping('forked/schedules')
    debug('end write to fork')

    const forkedSchedule = await blank.current('forked/schedules')
    // ? did it not update its parent ?
    const forkedPulseId = forkedSchedule.toString()
    debug('forkedPulseId', forkedPulseId)
    expect(forkedPulseId).not.toBe(originalPulseId)
  })
})
