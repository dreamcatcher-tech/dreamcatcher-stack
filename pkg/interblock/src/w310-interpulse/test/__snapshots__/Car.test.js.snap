// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Car export 1`] = `68`;
